examples/fishnet_api/main.py

```diff
@@ -97,43 +98,91 @@ async def index():
 
 
 @app.get("/datasets")
-async def datasets() -> List[Dataset]:
-    return await Dataset.fetch_all()
-
+async def datasets(view_as: Optional[str], by: Optional[str], page: Optional[int], page_size: Optional[int]) -> List[
+    Tuple[Dataset, Optional[DatasetPermissionStatus]]]:
+    """
+    Get all datasets.
 
-@app.get("/user/{address}/datasets")
-async def get_user_datasets(address: str) -> List[Dataset]:
-    return await Dataset.query(owner=address)
+    :param view_as: address of the user to view the datasets as and give additional permission information
+    :param by: address of the dataset owner to filter by
+    """
+    # TODO:
+    # - for the Owner (by) parameter:
+    #     - fetch all datasets for the owner
+    if not by:
+        HTTPException(status_code=404, detail="No Owner found")
+    ds_by_owner = await Dataset.where_eq(owner=by)
+    ts_ids = []
+    for rec in ds_by_owner:
+        ts_ids.append(rec.timeseriesIDs)
+
+    # convert into numpy array
+    ts_ids_np = np.array(ts_ids)
+    # convert the 2d array into 1d
+    ts_ids_lists = np.hstack(ts_ids_np)
+    # remove the duplicate values from the numpy array
+    ts_ids_unique = np.unique(ts_ids_lists)
+
+    # converting the numpy array into python normal list for pythonic calculation
+    # - for the Requestor (view_as) parameter:
+    #     - fetch all timeseries for each dataset,
+    ts_ids_lst = list(ts_ids_unique)
+
+    # This whole block should be executed for each fetched Dataset.Also, return the Dataset with it, not only the PermissionStatus.
+    dataset_by_requestor = await Dataset.where_eq(timeseriesIDs=ts_ids_lst)
+    #   - get all permissions for each timeseries & given requestor
+    permission_status = []
+    for rec in dataset_by_requestor:
+        permission_records = await Permission.where_eq(timeseriesID=rec.timeseriesIDs, requestor=view_as)
+        if not permission_records:
+            return [(rec, DatasetPermissionStatus.NOT_REQUESTED)]
+        for perm_rec in permission_records:
+            permission_status.append(perm_rec.status)
+        #   - respond with permission for dataset as approved, if all permissions are approved
+        if all(status == PermissionStatus.GRANTED for status in permission_status):
+            return [(rec, DatasetPermissionStatus.GRANTED)]
+            #     - respond with denied if at lest one is denied
+        elif PermissionStatus.DENIED in permission_status:
+            return [(rec, DatasetPermissionStatus.DENIED)]
+            #     - respond with pending if at least one is still pending
+        elif PermissionStatus.REQUESTED in permission_status:
+            return [(rec, DatasetPermissionStatus.REQUESTED)]
```

Comment:
This is not a .map() operation like in JavaScript. If you use a for loop to iterate over all datasets, you need to create another array, in which you put all the iteration results, which you are currently returning (as return value for the whole function).